Ogata text 5

ogwata opened the pull request:
ひきつづき、レビューをお願いします。

ogwata committed:
kobo-3のテスト結果を修正

ogwata committed:
box-shadow, border-imageの結果を修正

MurakamiShinyu commented:
"epubcsstest_voyager" → "epubcsstest_v1a" の変更がされてない箇所があります：

> - 1……テストファイルepubcsstest_voyagerによるテスト

ogwata committed:
hontoのSelectors Level 3を修正

ogwata committed:
epubcsstest_voyager→epubcsstest_v1a

ogwata commented:
> "epubcsstest_voyager" → "epubcsstest_v1a" の変更がされてない箇所があります：
> 
> > * 1……テストファイルepubcsstest_voyagerによるテスト

修正しました。

ogwata committed:
テスト結果に疑問があった項目も合わせて再テストしたEPUBリーダー

ogwata committed:
超縦書の記述を圧縮

ogwata committed:
表にキャプションを追加

ogwata committed:
BOOK☆WALKER→BOOK WALKER

ogwata committed:
kobo-3について追記

MurakamiShinyu commented:
## 「2-2-1 レイアウトエンジンと本テストとの関係」の「独自エンジン系」の説明

> 他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。これを嫌ってだろうか、コストをかけても独自のレイアウトエンジンを実装するEPUBリーダーも多い。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶ。

問題点：
- 「モダンブラウザー系」以外のものを一括して「独自エンジン系」と呼ぶことにしたということだが、それには古いブラウザエンジンを使っているものも含まれ、それらは「コストをかけても独自のレイアウトエンジンを実装」しているものではない。
  - 調査した12種類のEPUBリーダーのうち、Chromium・WebKit（古いバージョンを含む）に依存しない本当に独自といえそうなのは、Kindle、honto、Kinoppy、Romancerの4つだけであり、「多い」とはいえない。
  - ブラウザエンジンを利用しない理由が「コストをかけても独自のレイアウトエンジンを実装する」とは限らない。それぞれのベンダーの事情があるだろう。断定的に書くべきではない。

修正案：

> 他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。これを嫌ってコストをかけても独自のレイアウトエンジンを実装するEPUBリーダーもあるかもしれない。また、モダンブラウザー以外のレイアウトエンジンのライブラリを利用している可能性もある。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶ。

ogwata committed:
2-2-2削除

ogwata committed:
村上コメント対応

ogwata committed:
2-2-3のペーストミスを修正

ogwata committed:
テスト結果は2-2-3以降にゆずるが

MurakamiShinyu commented:
## OK・NGの誤りなど未修正なものまとめ

### Bibi

(Discardで報告: https://discord.com/channels/1011072514392330254/1052204695998955571/1225012573351448606 )

表2のBibiの
- 「outlineプロパティ」を「OK / NG」から「OK」に修正
  - 注1を削除
- 定義済みカウンタースタイル lower-greek についての注2を削除

表3のBibiの
- 「`@font-face`」を「OK / NG」から「NG」に修正
  - 注3を削除

### 超縦書

表3の超縦書の
- 「writing-modeプロパティ vertical-rl」を「OK / NG」から「NG」に修正

MurakamiShinyu commented:
> 前項でNGが多かったRomancerとKinoppy、およびhontoが、この項でもNGが多いのは変わらない。しかし、加えて前項ではOKばかりだったKindle、kobo-3、超縦書が、この項ではNGが多くなっている。これらは全て独自エンジン系のリーダーである。

「これらは全て独自エンジン系のリーダーである。」は削除したほうがよいと思います。
CSS変数をはじめ最近のCSSがサポートされているのをモダンブラウザ系それ以外を独自エンジン系と判定して分けたのだから、NGが多いのが「全て独自エンジン系のリーダーである」というのは当たり前のことなので。また、「独自エンジン系」というのは本当は独自ではない古いブラウザを使っているものも便宜上これに分類したものなので、あまり強調しない方がよいと思います。

ogwata committed:
表1でkoboを修正

ogwata committed:
kobo、BOOK☆WALKERの表記統一

ogwata committed:
表2からkobo-2削除

ogwata committed:
2-2-4表の注からバージョンを削除

ogwata committed:
表3からkobo-2削除、文字色を統一

MurakamiShinyu commented:
「2-1-3-1 掲載したテスト環境の一覧」で、超縦書のバージョンが「2.3.1」と「2.3.2」の２つあることに気がつきました。「2.3.2」のほうは間違いでしょう。（スプレッドシートでの誤操作？）

MurakamiShinyu commented:
「2-1-3 本調査におけるテスト環境の一覧」

> なお、背景色が緑色はモダンブラウザー系、水色は独自エンジン系である。これらの分類については2-2-1 レイアウトエンジンと本テストとの関係、及び2-2-2 MDN Web Docsの実装情報を活用を参照されたい。

「、及び2-2-2 MDN Web Docsの実装情報」は削除。

ogwata committed:
Bibiの結果を修正

ogwata commented:
ありがとうございます。
ここまでのコメントに対応しました。
ご確認ください。

ogwata committed:
村上コメント対応

MurakamiShinyu commented:
> 表2のBibiの
> 
> * 「outlineプロパティ」を「OK / NG」から「OK」に修正
>   * 注1を削除

注1がまだ削除されてません。

> 表3のBibiの
> 
> * 「`@font-face`」を「OK / NG」から「NG」に修正
>   * 注3を削除

注3がまだ削除されてません。

> 表3の超縦書の
> 
> * 「writing-modeプロパティ vertical-rl」を「OK / NG」から「NG」に修正

まだ修正されてません。

ogwata committed:
村上コメントのうち表2，表3のBibiだけ修正

ogwata committed:
表3凡例で注3を削除、以下注番号を順送りで詰める

MurakamiShinyu commented:
> 表3の超縦書の
> 
> * 「writing-modeプロパティ vertical-rl」を「OK / NG」から「NG」に修正

それから、この修正にともないその次の text-orientationのuprightとsidewaysの2行、「OK / N/A」を「N/A」に修正です。

ogwata committed:
表3の注番号、注4以降を順送りに詰める

ogwata committed:
超縦書OK / N/AをN/Aに

ogwata commented:
ここまで、修正できました。

ogwata committed:
2-2-3に画像を追加、.figure-bottom追加

ogwata committed:
図2画像追加

ogwata committed:
writing-mode-OK画像追加

ogwata committed:
画像位置調整

ogwata committed:
画像位置調整、キャプション

ogwata committed:
図2-図6入替

ogwata committed:
図2〜図5を元に戻す

ogwata committed:
figure要素入替

ogwata committed:
selector-3.pngを削除

ogwata committed:
Romancer→BinB、初出時に運営会社名

ogwata committed:
推敲

ogwata committed:
2-2-4を2つに分割

ogwata committed:
2-2-4-2最後まで

ogwata committed:
推敲

ogwata commented:
@MurakamiShinyu 

2-2-4まで書き終えたのでレビューをお願いします。

文章は終えたものの、レイアウトはまったく未完です。
原因はmain.cssの`.figure-right`、`.figure-bottom`がうまく動作せず、思い通りに図版がフロートしないところにあります。

https://github.com/jagat-xpub/viewer-test-2023/blob/0da95a64442dc5b07d96a56bab54b38a80896b63/css/main.css#L306C1-L346C1

希望としては、以下の2種のクラスがあるとうれしいです。

- 画像のトップがページエリアのトップに揃う
- 画像のボトムがページエリアのボトムに揃う

関連して、現在`notetext`が本文の当該箇所から次ページ以降にずれてしまっています。

これらの課題について教えていただけると大変ありがたいです。

ogwata committed:
OK、NGは鉤括弧で綴じる

ogwata committed:
推敲

ogwata committed:
項の多くのテスト結果も

ogwata committed:
「NG」が多いのが特徴だ

ogwata committed:
リーダー開発者名

ogwata committed:
独自な実装にある

ogwata committed:
テストファイルが違っても

ogwata committed:
制限している

ogwata committed:
iOS版に関しては

ogwata committed:
App Reviewガイドライン削除

ogwata committed:
App Reviewガイドライン復活

MurakamiShinyu commented on chap02.md:
```diff
@@ -167,11 +167,11 @@ lang: 'ja'
 ### 2-1-3 本調査におけるテスト環境の一覧
 
-今回の調査でテストしたEPUBリーダーの数は、全部で54にのぼる。そのうち、テスト結果を掲載したリーダーは36だ。その明細を<a href="#2-1-3-1-掲載したテスト環境の一覧">2-1-3-1 掲載したテスト環境の一覧</a>に掲げる。
+今回の調査でテストしたEPUBリーダーの数は、全部で55にのぼる。そのうち、テスト結果を掲載したリーダーは37だ。その明細を<a href="#2-1-3-1-掲載したテスト環境の一覧">2-1-3-1 掲載したテスト環境の一覧</a>に掲げる。
```
「EPUBリーダーの数は、全部で55」という表現だと、55もの異なるEPUBリーダーをテストしたかのようですが、そうではないので、「テスト環境の違いも入れると」を補った方がよいと思います。

```suggestion
今回の調査でテストしたEPUBリーダーの数は、テスト環境の違いも入れると、全部で55にのぼる。そのうち、テスト結果を掲載したものは37だ。その明細を<a href="#2-1-3-1-掲載したテスト環境の一覧">2-1-3-1 掲載したテスト環境の一覧</a>に掲げる。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -1514,27 +1427,32 @@ lang: 'ja'
-### 2-2-5 おもにモダンブラウザー系でサポートされるCSSモジュール
+#### 2-2-4-1　テスト結果の概要
+
+この項で報告するのは、11つのCSSモジュールにわたる54項目のテスト結果である（全82項目の約66%）。ここで言う「おもにモダンブラウザー系でサポートされるSSモジュール」とは、簡単に言うと前項で掲載したテスト項目以外全てである。言い換えると、「モダンブラウザー系だけでなく、独自エンジン系の一部でもサポートされるCSSモジュール」とも言える。では**表3**のテスト結果を見ていただこう。
```
- 「11つの」→「11の」
- 「SSモジュール」→「CSSモジュール」

```suggestion
この項で報告するのは、11のCSSモジュールにわたる54項目のテスト結果である（全82項目の約66%）。ここで言う「おもにモダンブラウザー系でサポートされるCSSモジュール」とは、簡単に言うと前項で掲載したテスト項目以外全てである。言い換えると、「モダンブラウザー系だけでなく、独自エンジン系の一部でもサポートされるCSSモジュール」とも言える。では**表3**のテスト結果を見ていただこう。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,33 +2567,97 @@ lang: 'ja'
+- 5……`vertical-rl`がNGの場合、これを前提とするtext-orientationは無効（N/A）
+- 6……掲載した全てのテスト環境で2段組がNGの場合、これを前提とする他のプロパティは無効（N/A）
+- 7……掲載した全てのテスト環境でmix-blend-modeがNGの場合、これを前提とするisolationは無効（N/A）
+- 8……Windows上のKindle 2.3.0のみNG。他はすべてOK
```
「Windows上のKindle 2.3.0のみNG」と細かなバージョンを書いているのは、Windows上のKindleには、Kindle for PCのほかにKindle Previewerもあるからかと思います。しかし、「Kindle 2.3.0のみNG」だとほかのバージョン（2.3.1など）だとOKなのかと誤解されるかもしれません。そこで「Windows上のKindle for PCのみNG」とするとよいかと思います。

```suggestion
- 8……Windows上のKindle for PCのみNG。他はすべてOK
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,33 +2567,97 @@ lang: 'ja'
+  <figcaption><strong>図1</strong> 左からKindle（Windows版 v2.3.1(70682)）、kobo-other (Mac版 v4.37.17113)、超縦書（v2.3.1）。</figcaption>
+</figure>
+
+すべてのテスト項目を説明するには紙幅が足りないので、ここまでとくに拘ってきた「writing-mode プロパティ vertical-rl」を例に挙げて、スクリーンショットとともに見ていこう。
```
- 「ここまでとくに拘ってきた」では拘わる意味がわかりません。「日本語のEPUBで特に重要な」などと直すとよいと思います。

```suggestion
すべてのテスト項目を説明するには紙幅が足りないので、日本語のEPUBで特に重要な「writing-mode プロパティ vertical-rl」を例に挙げて、スクリーンショットとともに見ていこう。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,33 +2567,97 @@ lang: 'ja'
+
+なお、テストから本稿執筆まで間隔が空いてしまったので、スクリーンショットは執筆時点（2024年4月11日）の最新版で撮り直した。リーダーのバージョンがテスト時と違うものはいちいちコメントしたが、テスト結果に変わりはなかった。なおまた、テストファイルはBinBだけがepubcsstest_v1a.epubで、他の全てはepubcsstest_v2.epubである。ただし「writing-mode プロパティ」に関してはテストファイルが違ってもテスト内容に基本的な違いはないので、結果が変わることは考えにくい。
+
+まず「NG」だったものから紹介すると、**図1**左からKindleのうちWindows（テスト時v2.3.0 (70673)→v2.3.1(70682)）、kobo-other（スクリーンショットはMac版だが、Windows版も同じ結果。v4.37.17113）、そして超縦書（Windows版 v2.3.1）。この3つのリーダーは`writing-mode: vertical-rl;`を指定しても縦書きで表示しなかった。これらのリーダーで縦書きを表示したい場合は、`-epub-`などの接頭辞をつける必要がある。
```
EPUB仕様では`-epub-`以外の接頭辞を認めてないので、「`-epub-`などの接頭辞」の「などの」は入れないほうがよいでしょう。

```suggestion
まず「NG」だったものから紹介すると、**図1**左からKindleのうちWindows（テスト時v2.3.0 (70673)→v2.3.1(70682)）、kobo-other（スクリーンショットはMac版だが、Windows版も同じ結果。v4.37.17113）、そして超縦書（Windows版 v2.3.1）。この3つのリーダーは`writing-mode: vertical-rl;`を指定しても縦書きで表示しなかった。これらのリーダーで縦書きを表示したい場合は、`-epub-`接頭辞をつける必要がある。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,33 +2567,97 @@ lang: 'ja'
+
+他方、Kindle（専用端末、モバイル端末、デスクトップ）と[Kindle Previewer](https://docs.google.com/spreadsheets/d/1xKDlL4TrMHMa1qq2QsWcXLEGMPjx-JWcTdaw_8KkftE/edit?pli=1#gid=0&range=P2:Q6)は、それぞれCSSモジュールの対応程度が異なる。しかしKindleの中では、CSSモジュールの対応程度はハードウェアやOSの違いが少ない。つまり、Kindleの中ではいくつかの例外を除きハードウェアやOSの違いに関わらず「OK」と「NG」がよく揃っていると言える。ただし、ここでも「writing-mode プロパティ vertical-rl」はその例外で、Windows版のKindleだけが「NG」だった（図1）。
+
+koboについても触れておこう。同じkoboでありながらkobo-otherが「NG」（図1）、kobo-mobileが「OK」（図2）と結果が分かれていることに興味が引かれるかもしれない。しかし、前者は独自エンジン系、後者はモダンブラウザー系なので、じつはレイアウトエンジンの違いが出た、分かりやすい結果とも言える。
```
「しかし、前者は独自エンジン系、後者はモダンブラウザー系なので、じつはレイアウトエンジンの違いが出た、分かりやすい結果とも言える」という書き方は少し違和感があります。テスト結果が違っていることから、独自エンジン系、モダンブラウザー系と分けたのであって、あらかじめそのように分けていたわけではありません。また、kobo-otherは実際には独自なエンジンではなくて古いブラウザエンジンです。



```suggestion
koboについても触れておこう。同じkoboでありながらkobo-otherが「NG」（図1）、kobo-mobileが「OK」（図2）と結果が分かれていることに興味が引かれるかもしれない。これは、前者は独自エンジン系（モダンではないブラウザーエンジン利用を含む）、後者はモダンブラウザー系というレイアウトエンジンの違いが出た、分かりやすい結果とも言える。
```

ogwata committed:
Update chap02.md

テスト環境の違いも入れると

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
Update chap02.md

11つの

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
Update chap02.md

Windows上のKindle for PCのみNG

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
Update chap02.md

日本語のEPUBで特に重要な

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
Update chap02.md

`-epub-`接頭辞をつける必要がある

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
Update chap02.md

というレイアウトエンジンの違いが出た、分かりやすい結果

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
小形克宏／村上真雄

ogwata committed:
ブックの結果の村上コメントに対応

ogwata committed:
2-3途中

ogwata committed:
2-3最後まで

ogwata committed:
国際規格を国内で効率よく運用するため

ogwata committed:
「電書協EPUB 3 制作ガイド ver.1.0」を公開します

MurakamiShinyu commented:
> このとき前述2項目と、以前からテスト結果に疑問があった項目も合わせて再テストしたEPUBリーダーのうち、誤判定がみつかった（言い換えれば`visibility`をサポートしていなかった）ものは、以下の11のリーダーであった。

これはだいぶ間違ってます。誤判定の原因となったのは `visibility` だけではありません。たとえば長さの単位のテストの誤判定にはvisibilityは関係ありません。
`visibility` が誤判定の原因になったのは Kinoppy と BinB および KindleのO列（Kindle for Mac 1.40.1）だけです。それ以外は `visibility` 以外の問題での誤判定、あるいは、誤判定かどうかが分からなかったため、確認のため再テストしたというものです。


次のように修正することを提案します：

- 「ここで使ったテストファイルの当初のバージョンでは、多くのテスト項目においてCSS レベル2.1のプロパティである`visibility`を使っていた。」の前に、「一例をあげると、」を入れる。
- 「このとき前述2項目と、以前からテスト結果に疑問があった項目も合わせて再テストしたEPUBリーダーのうち、誤判定がみつかった（言い換えれば`visibility`をサポートしていなかった）ものは、以下の11のリーダーであった。」とそのリーダーのリストを削除。

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
-- 8……掲載した全てのテスト環境でmix-blend-modeがNG。これを前提とするisolationは無効（N/A）
-- 9……Windows上のKindle 2.3.0のみNG。他はすべてOK
+- 1……iOSがOK、AndroidがNG
+- 2……一部のMac環境、及びiOSでNG
```
iPhoneでの「NG」は誤りだったことが分かったので、「及びiOS」は削除

```suggestion
- 2……一部のMac環境でNG
```

himorin commented:
> > このとき前述2項目と、以前からテスト結果に疑問があった項目も合わせて再テストしたEPUBリーダーのうち、誤判定がみつかった（言い換えれば`visibility`をサポートしていなかった）ものは、以下の11のリーダーであった。
> 
> これはだいぶ間違ってます。誤判定の原因となったのは `visibility` だけではありません。たとえば長さの単位のテストの誤判定にはvisibilityは関係ありません。 `visibility` が誤判定の原因になったのは Kinoppy と BinB および KindleのO列（Kindle for Mac 1.40.1）だけです。それ以外は `visibility` 以外の問題での誤判定、あるいは、誤判定かどうかが分からなかったため、確認のため再テストしたというものです。

これの関係で、一番初めの頃にもちとちゃんと指摘しておくべきだったところなんですが、テストの依存関係ってなんかデータ作られたりしてます？たとえば、XXのテストは`visibility`に依存している、とか。
web-platform-testsではそんな感じの指定をするようなプロパティーがあった気がする(気のせいかも・・・)ので、何らかの依存関係を一覧するものを作ってもいいのかな、とは思ってます。どこまでリストするかという問題はあるかもしれませんが。。。

ogwata committed:
「及びiOS」は削除

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

MurakamiShinyu commented:
上記、「次のように修正することを提案します」の2項目目の引用がコピペミスで間違ってた（1項目目と同じになってた）のを直しました。

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+
+本章冒頭<a href="#2-1-はじめに">「はじめに」</a>で、EPUBの現行規格である3.3は、[“CSS Snapshot”]((https://www.w3.org/TR/CSS/)) のうち、「CSSの公式的な定義に含まれるCSSモジュール」、つまり最も実装が安定している（普及している）CSSモジュールのサポートを求めていることを説明した。2種のうちモダンブラウザー系のEPUBリーダーは、この要件を満たしていると言ってよい。
+
+一方、独自エンジン系をCSSモジュールのサポートという観点から見ると、厳しい見方をせざるを得ない。EPUBリーダーは、現行規格が求める “CSS Snapshot” はもちろん、前バージョンであるEPUB 3.0.1が求めるCSSモジュールの実装要件 “CSS Profile”（[原文](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-profile)、[和訳](https://imagedrive.github.io/spec/epub301-contentdocs.xhtml#sec-css)）も、そのまた前のバージョンEPUB 3.0の “CSS Profile”（[原文](https://idpf.org/epub/30/spec/epub30-contentdocs-20111011.h ...
```
- 「そのまた前のバージョンEPUB 3.0の…ですら」まで書くのはしつこすぎるし不要だろう。
- すでにコメントしたとおり、visibilityが問題なのは独自エンジン系の一部でしかないので、修正

```suggestion
一方、独自エンジン系をCSSモジュールのサポートという観点から見ると、厳しい見方をせざるを得ない。EPUBリーダーは、現行規格が求める “CSS Snapshot” はもちろん、前バージョンであるEPUB 3.0.1が求めるCSSモジュールの実装要件 “CSS Profile”（[原文](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-profile)、[和訳](https://imagedrive.github.io/spec/epub301-contentdocs.xhtml#sec-css)）もサポートしているか怪しい。というのは、<a href="#2-2-4-メジャーなepubリーダーでサポートされるcssモジュール">2-2-4</a>で説明したように、いくつかの独自エンジン系のリーダーでCSS レベル2.1のプロパティである[visibility](https://www.w3.org/TR/2011/REC-CSS2-20110607/visufx.html#visibility)などをサポートしていないことが分かったからだ。前述 “CSS Profile” ではCSS レベル2.1のフルサポートが求められている。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+
+それにしても、なぜ独自エンジン系はこれほどCSSモジュールのサポートに消極的なのか、そのような素朴な疑問も浮かんでくる。本来この種の調査報告で推測など慎むべきだが、テストの結果得られた「仮説」として述べておきたい。
+
+独自エンジン系のEPUBリーダーは、EPUBではなく[電書協ガイド](http://ebpaj.jp/counsel/guide)のサポートを目指したのではないだろうか。
```
「EPUBではなく」は余計でしょう。EPUB仕様のサポートを目指すが、独自実装でEPUB3 CSS Profileをフルにサポートするのは、膨大な開発コストがかかり現実的ではないため、サポート範囲を限定するのは開発者にとっては当然のことです。そして日本語EPUBで要求されていることは電書協ガイドにまとめられているのだから、それを満足することを目指すのはとても自然なことです。

```suggestion
独自エンジン系のEPUBリーダーは、[電書協ガイド](http://ebpaj.jp/counsel/guide)のサポートを目指したのではないだろうか。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+この電書協ガイドとは、電子書籍の出版社による業界団体、日本電子書籍出版社協会（現在は[デジタル出版者連盟](http://ebpaj.jp/)）によって、「加盟社のための一般書のEPUB 3を制作する際のガイド」として2012年にver1.0が公開された規約だ<span class="notetext">[「電書協EPUB 3 制作ガイド ver.1.0」を公開します](http://www.ebpaj.jp/guide/guide.html)</span>
+。制定当時はEPUB 3.0に対応していたが、現在はEPUB 3.0.1（2014年）に対応した[ver.1.1.3 / 2015年1月1日版](http://www.idpf.org/epub/301/spec/epub-publications-20140626.html)にアップデートしている。
+
+このver.1.1.3をダウンロードすると「20150101-CSS機能一覧.pdf」という文書があり、そこに利用可能なCSS機能がリストアップされている。そのほとんどを占めるプロパティは、CSS レベル2.1のサブセットだが、以下のようなCSS レベル3のプロパティもふくまれている（一部は当該仕様が当時勧告前だったことから接頭辞を付けるよう指示）。
```
- 「以下のようなCSS レベル3のプロパティもふくまれている（一部は当該仕様が当時勧告前だったことから接頭辞を付けるよう指示）。」は、「以下のようなレベル3のプロパティも`-epub-`接頭辞つきでふくまれている。」に直します。「CSS レベル3」という表現は不正確（CSS全体でなく各モジュールごとのレベルなので）なのでここは単に「レベル3」のほうがよいです。ここにあげたレベル3のプロパティは一部ではなくみな接頭辞が必要なので「`-epub-`接頭辞つきで」という書き方にします。（電書協ガイドには`-webkit-`も載ってますがそれは無視してよいでしょう）

```suggestion
このver.1.1.3をダウンロードすると「20150101-CSS機能一覧.pdf」という文書があり、そこに利用可能なCSS機能がリストアップされている。そのほとんどを占めるプロパティは、CSS レベル2.1のサブセットだが、以下のようなレベル3のプロパティも`-epub-`接頭辞つきでふくまれている。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+
+1. CSS Writing Modes Level 3（writing-mode プロパティ vertical-rl、text-combine-upright text-orientation）
+2. CSS Text Decoration Module Level 3 （text-emphasis-style）
+3. CSS Text（text-align、line-break）
```
- 「`-epub-`接頭辞つきでふくまれている」と直したので、そのリストにのせるプロパティも接頭辞付きにします。
- `text-combine-upright` は古い仕様の接頭辞付きの `-epub-text-combine` しか独自系リーダーでは使えないのでこれを載せます。
- 「CSS Text（text-align、line-break）」はその上の「CSS Text Decoration Module Level 3」にあわせて「Module Level 3」を入れます。それから、text-alignはCSSレベル2.1のプロパティなのでここには載せません。

```suggestion
1. CSS Writing Modes Level 3（-epub-writing-mode、-epub-text-combine、-epub-text-orientation）
2. CSS Text Decoration Module Level 3 （-epub-text-emphasis-style）
3. CSS Text Module Level 3（-epub-line-break）
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+
+本テストでは、<a href="#2-1-2-テスト結果を掲載したcssモジュール">2-1-2</a>、及び<a href="#2-2-1-レイアウトエンジンと本テストとの関係">2-2-1</a>でも述べたように、まさかCSS レベル2.1をフルサポートしないEPUBリーダーがあるとは考えず、これをテスト項目から外していた。また、上記1〜3についても同じ理由でテスト項目に入れていたのは1だけだった。
+
+したがって、電書協ガイドver.1.1.3が求める「CSS レベル2.1のサブセットと上記1〜3」のほとんどが本テストの対象外だ。その上での仮説だが、独自エンジン系のリーダーがこの「20150101-CSS機能一覧.pdf」のサポートだけでよいと割り切った実装をしたとすれば、最新のCSSモジュールのサポートに消極的だった理由として説明がつくのではないか。
```
- 「上記1〜3」を「`-epub-`接頭辞つきプロパティ」に直します。
- 「まさか」は強すぎるので削除します。「CSS レベル2.1をフルサポートしないEPUBリーダー」を想定しなかったのは、テストを作成した私の落ち度です。想定して当然のものでした。
  - ちなみに、EPUB 2.0 ではCSS 2.1のうちサポートが必須なサブセットが規定されていました。それに このテストで問題になったvisibilityやoverflowプロパティは含まれていません。参考：https://lostandfound.github.io/epub2spec-ja/ops/ops_2.0_final_spec_ja.html#Section1.3.5
- 接頭辞付きのプロパティはもともと本テストの対象外なので「また、上記1〜3についても同じ理由でテスト項目に入れていたのは1だけだった。」は削除。

```suggestion
ここまでをまとめると、電書協ガイドver.1.1.3で使用されているCSS機能は、CSS レベル2.1のサブセットと`-epub-`接頭辞つきプロパティである。

本テストでは、<a href="#2-1-2-テスト結果を掲載したcssモジュール">2-1-2</a>、及び<a href="#2-2-1-レイアウトエンジンと本テストとの関係">2-2-1</a>でも述べたように、CSS レベル2.1をフルサポートしないEPUBリーダーがあるとは考えず、これをテスト項目から外していた。

したがって、電書協ガイドver.1.1.3が求める「CSS レベル2.1のサブセットと`-epub-`接頭辞つきプロパティ」は本テストの対象外だ。その上での仮説だが、独自エンジン系のリーダーがこの「20150101-CSS機能一覧.pdf」のサポートだけでよいと割り切った実装をしたとすれば、最新のCSSモジュールのサポートに消極的だった理由として説明がつくのではないか。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+
+したがって、電書協ガイドver.1.1.3が求める「CSS レベル2.1のサブセットと上記1〜3」のほとんどが本テストの対象外だ。その上での仮説だが、独自エンジン系のリーダーがこの「20150101-CSS機能一覧.pdf」のサポートだけでよいと割り切った実装をしたとすれば、最新のCSSモジュールのサポートに消極的だった理由として説明がつくのではないか。
+
+たとえば、<a href="#2-2-4-2writing-mode-プロパティ-vertical-rlに絞ったテスト結果">2-1-2</a>、及び<a href="#2-2-1-レイアウトエンジンと本テストとの関係">2-2-4-2</a>で述べたように、Kindle（Windows版）、kobo-other (Mac版／Windows版)、超縦書の4つはwriting-mode プロパティは接頭辞をつける必要がある。これはまさに電書協ガイドver.1.1.3の通りなのだ。では、これら以外の独自エンジン系リーダーは、なぜ接頭辞をつけないよう修正したのかといえば、仕様が勧告になったのでその必要がなくなったと考えるのが自然だろう。むしろ勧告後も接頭辞付きのルールを変えない電書協ガイドの方が不自然と言える。
```
- 「むしろ勧告後も接頭辞付きのルールを変えない電書協ガイドの方が不自然と言える。」は良いが、次のように補足を加えるとよいのではないか：

  > 互換性のために`-epub-`接頭辞つきプロパティを残すことは必要であるにしても、現在の標準である接頭辞なしのプロパティと併記するべきではないか。

MurakamiShinyu commented on chap02.md:
```diff
@@ -2718,34 +2579,130 @@ lang: 'ja'
+
+たとえば、<a href="#2-2-4-2writing-mode-プロパティ-vertical-rlに絞ったテスト結果">2-1-2</a>、及び<a href="#2-2-1-レイアウトエンジンと本テストとの関係">2-2-4-2</a>で述べたように、Kindle（Windows版）、kobo-other (Mac版／Windows版)、超縦書の4つはwriting-mode プロパティは接頭辞をつける必要がある。これはまさに電書協ガイドver.1.1.3の通りなのだ。では、これら以外の独自エンジン系リーダーは、なぜ接頭辞をつけないよう修正したのかといえば、仕様が勧告になったのでその必要がなくなったと考えるのが自然だろう。むしろ勧告後も接頭辞付きのルールを変えない電書協ガイドの方が不自然と言える。
+
+ただし、この「CSS レベル2.1のサブセットと上記1〜3」というのは、あくまでもEPUB制作者の便を図るための国内ルールであることを忘れてはいけないだろう。国際規格を国内で効率よく運用するためサブセットを規定することはなんの問題もない。もっとも、それをリーダーが実装するとなると別次元の問題だ。万が一、電書協ガイドを実装したリーダーがあったとして、はたしてそれを「EPUB」リーダーと呼んでよいのか、いささか疑問が残るところではある。
```
コメントしたとおり独自実装で「CSS レベル2.1のサブセットと`-epub-`接頭辞つきプロパティ」しかサポートされてないのは仕方のないところなので、そのことを責めるような終わり方にはしないほうがよいと思います。

```suggestion
```

ogwata replied:
「ただし」以下の段落を削除というサジェストですが、これはこの段落の再考を促すものと受け取りました。考え直すためにいくつか質問をさせてください。

1. EPUB 3 CSS Profileの規定に反したリーダーは、EPUB 3.0.1に不適合ということになりませんか？
2. そもそも電書協ガイドver.1.1.3は、EPUB 3.0.1に適合しているのでしょうか？

EPUB 3.0.1では、3.1に以下のような規定があります。

> 3.1 Content Conformance
> A conformant [EPUB Style Sheet](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#gloss-stylesheet) must meet all of the following criteria:
> 
> [›](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#confreq-content-css-support) It must adhere to all content restrictions given in [EPUB 3 CSS Profile](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-profile).

この「It must adhere to all content restrictions given in EPUB 3 CSS Profile.」というのが私には難解ですが、要するに「EPUB 3 CSS Profileの規定に従うこと（従わなければ不適合）」ということだと思います。

そうすると、たとえばEPUB 3 CSS Profile [3.3.4](https://idpf.org/epub/30/spec/epub30-contentdocs-20111011.html#sec-css-fonts)において、「MUST」で規定されているunicode-rangeに関して、本テストで「NG」だったKindle、ブック6.1、honto、Kinoppy、BinB、Bibiは、EPUB 3.0.1に不適合ということにならないでしょうか？（ここにモダンブラウザー系が含まれているので、論旨を変更しないと……）

つぎに、この「unicode-range」がよい例なのですが、電書協ガイドver.1.1.3「20150101-CSS機能一覧.pdf」にはunicode-rangeは触れられていないのです。つまり、電書協ガイドver.1.1.3はEPUB 3.0.1の規定に反しているともとれますが、これは私の誤解でしょうか？

ここまで、細かな規格論を書きましたが、テスト項目1つがNGだからと言って、大幅に書き直さねばとまで思いません。ここで私が知りたいのは、EPUB仕様とそのサブセットであるところの電書協ガイドの関係を、どう考えればよいのか、サブセットに適合しただけで親仕様の適合を名乗ってよいのかということです。

私自身は、原稿に書いたように、「制作者がサブセットでコンテンツを作るのはOK（適合）、リーダーの実装者が電書協ガイドを実装するのは自由だが、その実装をEPUBリーダーと名乗るのは非適合」と考えています。

村上さんの答えを手がかりにして、結論部分を再考したいと考えています。
![IMG_0985](https://github.com/jagat-xpub/viewer-test-2023/assets/31482860/29099f61-b8c2-41c9-b37d-a1cf10a6b59e)

MurakamiShinyu replied:
> EPUB 3.0.1では、3.1に以下のような規定があります。
> 
> > 3.1 Content Conformance
> > A conformant [EPUB Style Sheet](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#gloss-stylesheet) must meet all of the following criteria:
> > 
> > [›](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#confreq-content-css-support) It must adhere to all content restrictions given in [EPUB 3 CSS Profile](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-profile).
> 
> この「It must adhere to all content restrictions given in EPUB 3 CSS Profile.」というのが私には難解ですが、要するに「EPUB 3 CSS Profileの規定に従うこと（従わなければ不適合）」ということだと思います。

誤解があります。

まず、このセクション「3.1 Content Conformance」は、EPUBコンテンツ制作者が用意するスタイルシートに対しての規定であり、EPUBリーダーに対しての規定ではありません。

そして、「It must adhere to all content restrictions given in EPUB 3 CSS Profile.」（EPUB 3 CSS Profileで与えられるコンテンツ制限に従う必要がある）というのは、要するにEPUB 3 CSS Profileに含まれるCSS仕様にある規則に従ってスタイルシートを書くようにということでしょう。ここではCSSを書く決まりごと（構文とか使える文字とか）を守ることということで、CSSプロパティのサポートとは関係ありません。

EPUB 3 CSS Profileで必須(MUST)とされているCSSの機能はとても限られています。CSS Profileに含まれる機能すべてをサポートしなければならないとは書かれていません。そこが誤解されているところです。

CSS 2.1 (https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-2.1) については、MUSTと書かれているのは「Reading Systems that have a CSS Viewport MUST support the font-family property.」（CSSビューポートを持つリーディングシステムは、font-familyプロパティをサポートしなければならない）だけです。visibilityプロパティなどはMUSTではありません。

CSS 2.1 の font-family プロパティ以外でMUSTと書かれているのは、
「Reading Systems with a CSS Viewport must support OpenType and WOFF fonts embedded using the `@font-face` rule.」（CSSビューポートを持つリーディングシステムは、`@font-face`ルールを使って埋め込まれたOpenTypeおよび WOFF フォントをサポートしなければならない）と
「Reading Systems must support at least the following `@font-face` font descriptors.」（リーディングシステムは、少なくとも次の`@font-face`フォント記述子をサポートしなければならない）です。

この`@font-face`ルールや`unicode-range`記述子のサポートが必須とされているところだけは確かに厳しい（この規定を満たせないEPUBリーダーが多い）のですが、そこには
「For forwards compatibility with EPUB 2 Reading Systems that do not support `@font-face` rules, authors should reference a generic font using the font-family property.」（`@font-face`ルールをサポートしていないEPUB 2リーディングシステムとの互換性のために、著者はfont-familyプロパティを使ってジェネリックフォントを参照すべきです）とも書かれています。

> 
> そうすると、たとえばEPUB 3 CSS Profile [3.3.4](https://idpf.org/epub/30/spec/epub30-contentdocs-20111011.html#sec-css-fonts)において、「MUST」で規定されているunicode-rangeに関して、本テストで「NG」だったKindle、ブック6.1、honto、Kinoppy、BinB、Bibiは、EPUB 3.0.1に不適合ということにならないでしょうか？（ここにモダンブラウザー系が含まれているので、論旨を変更しないと……）

まず、「ブック6.1」での`@font-face`ルールや`unicode-range`記述子が「NG」というのは、調べないと理由が不明です（ブックで使われるWebKitは当然それらをサポートしてます）。私のMacで確認した「ブック3.2」と「ブック6.3」では「OK」です。特定の環境で不具合が起きているのかもしれません。

「…は、EPUB 3.0.1に不適合ということにならないでしょうか？」ですが、EPUB3やCSSレベル2のように大きな仕様の場合、完全に仕様通りの実装というのはとても難しく、実装によって不完全さや制限があるのは普通のことです。100点満点の実装というのはまず存在しません。`@font-face`ルールが機能しないだけで「EPUB 3.0.1に不適合」というのは言い過ぎでしょう。

> 
> つぎに、この「unicode-range」がよい例なのですが、電書協ガイドver.1.1.3「20150101-CSS機能一覧.pdf」にはunicode-rangeは触れられていないのです。つまり、電書協ガイドver.1.1.3はEPUB 3.0.1の規定に反しているともとれますが、これは私の誤解でしょうか？

誤解です。電書協ガイドは、EPUB制作者のためのガイドです。EPUBリーダーの実装についてのガイドではありません。unicode-rangeは多くのEPUBリーダーで使えないのですから、EPUB制作者のためのガイドには載せないというのは当たり前のことです。

ogwata replied:
私の誤解を正してくれて、ありがとうございます。納得いたしました。

「制作者がサブセットでコンテンツを作るのはOK（適合）、リーダーの実装者が電書協ガイドを実装するのは自由だが、その実装をEPUBリーダーと名乗るのは非適合」についてはいかがでしょう。これも誤解でしょうか？

もしもこれが正しい場合、現在の2705行目は内容としては問題はなく、責めるように受け取られるところが問題、ということになります。

MurakamiShinyu replied:
> 「制作者がサブセットでコンテンツを作るのはOK（適合）、リーダーの実装者が電書協ガイドを実装するのは自由だが、その実装をEPUBリーダーと名乗るのは非適合」についてはいかがでしょう。これも誤解でしょうか？

はい、誤解です。

「電書協ガイド」はEPUB制作者のためのガイドなので、「リーダーの実装者が電書協ガイドを実装する」は正しくは「リーダーの実装者が電書協ガイドに記載されている範囲のCSSの機能を実装する」ということかと思います。もちろんそれは自由で、日本語EPUBのリーダーの実装者にとって合理的な判断でしょう。それが「EPUBリーダーと名乗るのは非適合」かですが、そのようなことはありません。

EPUB3の現在の仕様である EPUB 3.3 に、CSSのサポートが不十分だからといって非適合とするような記述はありません。そのことは、EPUB 3.3 (W3C Recommendation, 2023) とその前の EPUB 3.2 (Final Community Group Specification, 2019) を比べてみるとわかります。EPUB 3.2 では厳しい要求があったのが、EPUB 3.3 では緩和されています。EPUB 3.2のCommunity Group仕様から正式なW3C勧告に格上げされるときに、W3C勧告は複数の実装があることなどが必要であるために厳しい要求が緩和されたと考えられます。そのEPUB 3.2にあった厳しい要求は次のものでした：

EPUB 3.2 の「4. CSS Style Sheets」の「4.3 Reading System Conformance」というセクション：
https://www.w3.org/publishing/epub32/epub-contentdocs.html#sec-css-rs-conf

> A conformant EPUB Reading System MUST meet all of the following criteria for processing CSS Style Sheets:
> 
> -  It MUST support the official definition of CSS as described in the [CSSSnapshot].
> -  It SHOULD support all applicable modules in [CSSSnapshot] that have reached at least Candidate Recommendation status [W3CProcess] (and are widely implemented).
> -  It MUST support [TrueType], [OpenType], [WOFF] and [WOFF2] font resources referenced from `@font-face` rules.
> -  It MUST support all prefixed properties defined in CSS Style Sheets — Prefixed Properties.
> -  In addition to supporting CSS properties as defined above, its user agent style sheet SHOULD support the [HTML] suggested default rendering.
> -  It SHOULD respect Author CSS and user styles as defined in Reading System Overrides.
> 
> Reading System developers are encouraged to implement CSS support at the level of major browsers.

これでは、モダンブラウザ系のEPUBリーダーしか適合しないということになってしまいます。

独自エンジン系のEPUBリーダーにとって厳しすぎて無理、ということでこれではW3C勧告にはできないとなって当然でしょう。

そして、W3C勧告のEPUB 3.3では、この「4.3 Reading System Conformance」セクションが丸ごと削除されています。

EPUB 3.0.1 にあった`@font-face`ルールなどについてMUSTというのもなくなってます。

ogwata replied:
ご教示感謝です。
まとめますと、EPUB 3.3では「CSS Snapshotをサポートする」と書かれているものの、それはMUSTでも SHOULD でもMAYでもない（つまり空文）ということでしょうか？

もしそうでしたら、これは本テストの前提に関わることですね。「はじめに」を書いている時になにか居心地の悪さを感じていましたが、その時に言語化すべきでした。

MurakamiShinyu replied:
CSS SnapshotのCSSの公式の定義がサポートされることが望ましいけれど、どこまでサポートするかはEPUBリーダーの自由ということでしょう。

CSS Snapshotに次のように書かれてます：

- https://www.w3.org/TR/css-2023/#profiles

  > Not all implementations will implement all functionality defined in CSS.

  （すべての実装がCSSで定義されたすべての機能を実装するわけではない。）

- https://www.w3.org/TR/css-2023/#partial
  > 3.1. Partial Implementations

  …（部分的な実装を認めているということです。）

ogwata replied:
> …（部分的な実装を認めているということです。）

了解しました。その前提で最終段落を再考し、あわせて全体を見直します。

その上で、「部分実装を認められているCSS Snapshotへの対応をテストする意味」についても十分に考え直す必要があるかもしれません。

ogwata commented:
@MurakamiShinyu 

以下のコメントにもお答えいただけるとうれしいです。
https://github.com/jagat-xpub/viewer-test-2023/pull/5#issuecomment-2054074764

ogwata committed:
村上コメント対応

ogwata committed:
「再テストが必要だったことで」他をトル

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
「EPUBではなく」をトル

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata commented:
> > このとき前述2項目と、以前からテスト結果に疑問があった項目も合わせて再テストしたEPUBリーダーのうち、誤判定がみつかった（言い換えれば`visibility`をサポートしていなかった）ものは、以下の11のリーダーであった。
> 
> これはだいぶ間違ってます。誤判定の原因となったのは `visibility` だけではありません。たとえば長さの単位のテストの誤判定にはvisibilityは関係ありません。 `visibility` が誤判定の原因になったのは Kinoppy と BinB および KindleのO列（Kindle for Mac 1.40.1）だけです。それ以外は `visibility` 以外の問題での誤判定、あるいは、誤判定かどうかが分からなかったため、確認のため再テストしたというものです。
> 
> 次のように修正することを提案します：
> 
>     * 「ここで使ったテストファイルの当初のバージョンでは、多くのテスト項目においてCSS レベル2.1のプロパティである`visibility`を使っていた。」の前に、「一例をあげると、」を入れる。
> 
>     * 「このとき前述2項目と、以前からテスト結果に疑問があった項目も合わせて再テストしたEPUBリーダーのうち、誤判定がみつかった（言い換えれば`visibility`をサポートしていなかった）ものは、以下の11のリーダーであった。」とそのリーダーのリストを削除。

ひとまずご指示に従ってみます。

ogwata committed:
以下のようなレベル3のプロパティも`-epub-`接頭辞つきでふくまれている

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
リストのプロパティを接頭辞付きに他

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
CSS レベル2.1のサブセットと-epub-接頭辞つきプロパティ

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
接頭辞なしのプロパティと併記するべきではないか

ogwata committed:
接頭辞なしのプロパティと併記するべき

ogwata commented:
> これの関係で、一番初めの頃にもちとちゃんと指摘しておくべきだったところなんですが、テストの依存関係ってなんかデータ作られたりしてます？たとえば、XXのテストは`visibility`に依存している、とか。 web-platform-testsではそんな感じの指定をするようなプロパティーがあった気がする(気のせいかも・・・)ので、何らかの依存関係を一覧するものを作ってもいいのかな、とは思ってます。どこまでリストするかという問題はあるかもしれませんが。。。

ご指摘はありがたく、また私もテストの依存関係が明記されていた方がよいとは思うのですが、今となってはどうすることもできないというのが正直なところではないでしょうか。

MurakamiShinyu committed:
スタイル修正 .figure-bottom および .figure-top

MurakamiShinyu committed:
tableのマークアップの間違い（captionが最後にある）を修正

HTML仕様 4.9.1 The table element https://html.spec.whatwg.org/multipage/tables.html#the-table-element で
table要素の内容モデルは

> In this order: optionally a caption element, followed by zero or more colgroup elements, followed optionally by a thead element, followed by either zero or more tbody elements or one or more tr elements, followed optionally by a tfoot element, …

と書かれてます。captionがある場合、それは最初の要素である必要があります。
captionをtableの上に表示するか下に表示するかはCSSのcaption-sideプロパティで決まります。マークアップでのcaptionの位置は関係ありません。captionをtableの終わりに入れるマークアップでもChromium+Vivliostyleでたまたま動いてますが、HTMLとしては間違いです。

MurakamiShinyu commented:
> 文章は終えたものの、レイアウトはまったく未完です。 原因はmain.cssの`.figure-right`、`.figure-bottom`がうまく動作せず、思い通りに図版がフロートしないところにあります。
> 
> https://github.com/jagat-xpub/viewer-test-2023/blob/0da95a64442dc5b07d96a56bab54b38a80896b63/css/main.css#L306C1-L346C1
> 
> 希望としては、以下の2種のクラスがあるとうれしいです。
> 
> * 画像のトップがページエリアのトップに揃う
> * 画像のボトムがページエリアのボトムに揃う
> 
> 関連して、現在`notetext`が本文の当該箇所から次ページ以降にずれてしまっています。

脚注が本文の当該箇所から次ページ以降にずれてほとんど空白のページができてしまうことがあるのはVivliostyle.jsのバグのようで調べてます。ページフロート以外のフロート画像があるなどのときにこの現象が起きるようです。

コミット https://github.com/jagat-xpub/viewer-test-2023/pull/5/commits/a96c3bfdb89d18d19b7d099f199ad61d01459fc2 (スタイル修正 .figure-bottom および .figure-top)でご希望通りのページフロートに直しました。

MurakamiShinyu commented:
tableのマークアップの修正のコミット https://github.com/jagat-xpub/viewer-test-2023/pull/5/commits/2ab707a712f796510cc1a40a4bb23258ccdfc0b4 (tableのマークアップの間違い（captionが最後にある）を修正) をしました。

> tableのマークアップの間違い（captionが最後にある）を修正
>
> HTML仕様 4.9.1 The table element https://html.spec.whatwg.org/multipage/tables.html#the-table-element で
> table要素の内容モデルは
> 
> > In this order: optionally a caption element, followed by zero or more colgroup elements, followed optionally by a thead element, followed by either zero or more tbody elements or one or more tr elements, followed optionally by a tfoot element, …
> 
> と書かれてます。captionがある場合、それは最初の要素である必要があります。
> captionをtableの上に表示するか下に表示するかはCSSのcaption-sideプロパティで決まります。マークアップでのcaptionの位置は関係ありません。captionをtableの終わりに入れるマークアップでもChromium+Vivliostyleでたまたま動いてますが、HTMLとしては間違いです。
>

ogwata committed:
1-KindleiPhone-Kindlemac-koboiPhone差替

ogwata committed:
CSS SnapshotはMUSTではないと修正

ogwata committed:
推敲

MurakamiShinyu commented on chap02.md:
```diff
@@ -752,154 +762,94 @@ lang: 'ja'
 本稿では、現在使われている（つまりモダンな<span class="notetext">「モダン」の定義を補足すると、WebにおけるCSS仕様の使われ方が大きく変わったのは、2014年〜2016年に実装がはじまった[CSS変数（カスタムプロパティ）](https://developer.mozilla.org/ja/docs/Web/CSS/var#%E3%83%96%E3%83%A9%E3%82%A6%E3%82%B6%E3%83%BC%E3%81%AE%E4%BA%92%E6%8F%9B%E6%80%A7)だ。そこでこれが使えるかどうかを重要な指標とした。</span>）ブラウザーのレイアウトエンジンを利用したEPUBリーダーを、一括して**「モダンブラウザー系」**と呼ぶことにする。
 
-他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。これを嫌ってだろうか、コストをかけても独自のレイアウトエンジンを実装するEPUBリーダーも多い。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶ。
+他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。それ以外にも、過去の実装との互換性など、ベンダー固有の事情により独自のレイアウトエンジンを実装するEPUBリーダーもある。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶことにする。
```
モダンではないブラウザーのレイアウトエンジンを利用するEPUBリーダーについて触れるべきでしょう。それから、「過去の実装との互換性など、ベンダー固有の事情により」というのは推測でしかないので「…もある」と断定ではなく「…もあるようだ」と書くとよいと思います。

```suggestion
他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。それ以外にも、過去の実装との互換性など、ベンダー固有の事情により独自のレイアウトエンジン、あるいはモダンではないブラウザーのレイアウトエンジンを利用するEPUBリーダーもあるようだ。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶことにする。
```

ogwata replied:
以下にしようと思います。
あるようだ→あるだろう

ogwata replied:
「モダンではないブラウザーのレイアウトエンジンを利用する」は不正確でしょう。開発当初はモダンだったのですから。ちょっと考えますね。

ogwata replied:
以下でコミットしました。

他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。それ以外にも、過去の実装との互換性など、ベンダー固有の事情により独自のレイアウトエンジンを実装するEPUBリーダーもある。またブラウザーのレイアウトエンジンを実装したものの、時間が過ぎて「モダン」ではなくなったものもあるようだ。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶことにする。

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2583,108 @@ lang: 'ja'
+
+他方、Kindle（専用端末、モバイル端末、デスクトップ）と[Kindle Previewer](https://docs.google.com/spreadsheets/d/1xKDlL4TrMHMa1qq2QsWcXLEGMPjx-JWcTdaw_8KkftE/edit?pli=1#gid=0&range=P2:Q6)は、それぞれCSSモジュールの対応程度が異なる。しかしKindleの中では、CSSモジュールの対応程度はハードウェアやOSの違いが少ない。つまり、Kindleの中ではいくつかの例外を除きハードウェアやOSの違いに関わらず「OK」と「NG」がよく揃っていると言える。ただし、ここでも「writing-mode プロパティ vertical-rl」はその例外で、Windows版のKindleだけが「NG」だった（図1）。
+
+koboについても触れておこう。同じkoboでありながらkobo-otherが「NG」（図1）、kobo-mobileが「OK」（図2）と結果が分かれていることに興味が引かれるかもしれない。これは、前者は独自エンジン系（モダンではないレイアウトエンジン利用を含む）、後者はモダンブラウザー系というレイアウトエンジンの違いが出た、分かりやすい結果とも言える。
```
「モダンではないレイアウトエンジン」は、「モダンではないブラウザーのレイアウトエンジン」に直してください。「モダンではないレイアウトエンジン」ではそれがモダンではない独自の（既存ブラウザー以外の）レイアウトエンジンのことと読めますがそれは違うので。

```suggestion
koboについても触れておこう。同じkoboでありながらkobo-otherが「NG」（図1）、kobo-mobileが「OK」（図2）と結果が分かれていることに興味が引かれるかもしれない。これは、前者は独自エンジン系（モダンではないブラウザーのレイアウトエンジン利用を含む）、後者はモダンブラウザー系というレイアウトエンジンの違いが出た、分かりやすい結果とも言える。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2583,108 @@ lang: 'ja'
+
+koboについても触れておこう。同じkoboでありながらkobo-otherが「NG」（図1）、kobo-mobileが「OK」（図2）と結果が分かれていることに興味が引かれるかもしれない。これは、前者は独自エンジン系（モダンではないレイアウトエンジン利用を含む）、後者はモダンブラウザー系というレイアウトエンジンの違いが出た、分かりやすい結果とも言える。
+
+**図3**に移ろう。左からブック（Mac版、テスト時v6.1→v6.3(6040)、MURASAKI（Mac版、v2.4.1(30)）、honto（iPhone版 v6.62.0）。このうちブックとMURASAKIはモダンブラウザー系だ。残るhontoは独自エンジン系だが、「OK」の結果を出している。iOSでは「Webを閲覧するアプリ」<span class="notetext">[「App Reviewガイドライン」2.5.6（アップル）、最終更新日：2024年4月5日](https://developer.apple.com/jp/app-store/review/guidelines/#performance)</span>に関してはレイアウトエンジンをWebKitに制限している。EPUBリーダーの多くもiOS版に関してはWebKitを使う中で珍しい実装だ。
```
hontoのiOS版についての記述の「iOSでは「Webを閲覧するアプリ」［脚注：[「App Reviewガイドライン」2.5.6（アップル）、最終更新日：2024年4月5日](https://developer.apple.com/jp/app-store/review/guidelines/#performance)］に関してはレイアウトエンジンをWebKitに制限している。EPUBリーダーの多くもiOS版に関してはWebKitを使う中で珍しい実装だ。」は、次の理由により削除するほうがよいと思います：

- hontoについて「EPUBリーダーの多くもiOS版に関してはWebKitを使う中で珍しい実装」と書いてますが、それはKindleのiOS版にも該当するので「珍しい」といえません。
- hontoがWebKitを使っていないというのはたぶん不正確です。調べたところ（[「独自または古いブラウザエンジンを使うEPUBリーダーを調べた」の「honto」](https://gist.github.com/MurakamiShinyu/752b46ef8f758566b77932cb04079487#honto)）、ブラウザーエンジン上でJavaScriptなどで独自のレイアウトを行っているようです。したがって、iOS版ではWebKit上でそれが動いている可能性が高いです。
- 「iOSでは「Webを閲覧するアプリ」をWebKitに制限している」という情報は、よく知られていることだし、このWebKit制限とEPUBリーダーでの利用との関係はそう単純ではないので、書かない方がよけいな誤解を招かなくてよいと思います。

```suggestion
**図3**に移ろう。左からブック（Mac版、テスト時v6.1→v6.3(6040)、MURASAKI（Mac版、v2.4.1(30)）、honto（iPhone版 v6.62.0）。このうちブックとMURASAKIはモダンブラウザー系だ。残るhontoは独自エンジン系だが、「OK」の結果を出している。
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2583,108 @@ lang: 'ja'
+
+なお、hontoはMac版、Windows版も用意されているが、サイドロードに対応しているのはiOS、Android版だけであり、さらにサイドロードしたEPUBは同期の対象外であることから、Mac版、Windows版のテストはあきらめざるを得なかった。
+
+次に**図4**。左から、Book Walker（スクリーンショットはiPhone版だが、Android版も同じ結果。テスト時v7.4.7→v7.6.0）、Kinoppy（Windows版 テスト時v3.2.19→v3.3.20.329851E）、BinB（バージョン非表示）。このうちBook Walkerのみモダンブラウザー系で、KinoppyとBinBは独自エンジン系だ。このうち、BinBは<a href="#2-2-3-メジャーなepubリーダーでサポートされるcssモジュール">2-2-3</a>だけでなく、本項の多くのテスト結果も「NG」だったが、「writing-mode プロパティ vertical-rl」については「OK」としている。
```
「「OK」としている。」という表現は、何か問題があったけれども「OK」ということにした、という感じで、ひっかかります。「「OK」だった。」のほうが自然でないでしょうか？


```suggestion
次に**図4**。左から、Book Walker（スクリーンショットはiPhone版だが、Android版も同じ結果。テスト時v7.4.7→v7.6.0）、Kinoppy（Windows版 テスト時v3.2.19→v3.3.20.329851E）、BinB（バージョン非表示）。このうちBook Walkerのみモダンブラウザー系で、KinoppyとBinBは独自エンジン系だ。このうち、BinBは<a href="#2-2-3-メジャーなepubリーダーでサポートされるcssモジュール">2-2-3</a>だけでなく、本項の多くのテスト結果も「NG」だったが、「writing-mode プロパティ vertical-rl」については「OK」だった。
```

ogwata committed:
推敲

ogwata committed:
2-Book-MURASAKI-honto差替

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2583,122 @@ lang: 'ja'
+
+ここまでをまとめると、電書協ガイドver.1.1.3で使用されているCSS機能は、CSS レベル2.1のサブセットと`-epub-`接頭辞つきプロパティである。これを“CSS Snapshot” と比べると、ミニマムなセットと言えるだろう。
+
+本テストでは、<a href="#2-1-2-テスト結果を掲載したcssモジュール">2-1-2</a>、及び<a href="#2-2-1-レイアウトエンジンと本テストとの関係">2-2-1</a>でも述べたように、CSS レベル2.1をフルサポートしないEPUBリーダーがあることまで思いが及ばず、これをテスト項目から外していた。この点、テストファイルの開発方針として不適切だったことを率直に認めたい。
```
この部分同意できません。

CSS 2.1はとても大きな仕様なので、独自エンジンでそのフルサポートが難しいということは認識していました。私の落ち度はそのCSS 2.1の中でテストに利用したvisibilityやoverflowプロパティなどのサポートを前提としたテストを最初に作っていたことです。それについてはv1a以降で修正してます。

そしてその大きな仕様であるCSS 2.1をテスト項目に入れるとなると、膨大な作業量が発生します。
ですから、今回のテストは、EPUB3.0や電書協ガイドのころよりも新しくなっていると期待されるCSS機能に絞ってテストしたということです。私はそれがテストファイルの開発方針として不適切だったとは思いません。

```suggestion

```

ogwata replied:
了解です。削除しました。
ところで、反省点はなんでしょう？
不要であれば触れませんが。

MurakamiShinyu commented on chap02.md:
```diff
@@ -752,154 +762,94 @@ lang: 'ja'
+その上で、
+「CSSの公式的な定義に含まれるCSSモジュール」を、サポートするEPUBリーダーの数によって以下の2つのグループに分け、テスト結果を掲載する。
 
 1. <a href="#2-2-4-メジャーなepubリーダーでサポートされるcssモジュール">2-2-4 メジャーなEPUBリーダーでサポートされるCSSモジュール</a>
 2. <a href="#2-2-5-おもにモダンブラウザー系でサポートされるcssモジュール">2-2-5 おもにモダンブラウザー系でサポートされるCSSモジュール</a>
```
セクション番号の修正

```suggestion
1. <a href="#2-2-3-メジャーなepubリーダーでサポートされるcssモジュール">2-2-3 メジャーなEPUBリーダーでサポートされるCSSモジュール</a>
2. <a href="#2-2-4-おもにモダンブラウザー系でサポートされるcssモジュール">2-2-4 おもにモダンブラウザー系でサポートされるCSSモジュール</a>
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2583,122 @@ lang: 'ja'
+
+本章冒頭<a href="#2-1-はじめに">「はじめに」</a>で、EPUBの現行規格である3.3は、[“CSS Snapshot”]((https://www.w3.org/TR/CSS/)) のうち、「CSSの公式的な定義に含まれるCSSモジュール」、つまり最も実装が安定している（普及している）CSSモジュールのサポートを求めていることを説明した。2種のうちモダンブラウザー系のEPUBリーダーは、おおむねこの要件を満たしていると言ってよい。
+
+一方、独自エンジン系をCSSモジュールのサポートという観点から見ると、厳しい見方をせざるを得ない。EPUBリーダーは、現行規格が求める “CSS Snapshot” はもちろん、前バージョンであるEPUB 3.0.1が求めるCSSモジュールの実装要件 “CSS Profile”（[原文](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-profile)、[和訳](https://imagedrive.github.io/spec/epub301-contentdocs.xhtml#sec-css)）もサポートしているか怪しい。というのは、<a href="#2-2-4-メジャーなepubリーダーでサポートされるcssモジュール">2-2-4</a>で説明したように、いくつかの独自エンジン系のリーダ ...
```
セクション番号の修正

```suggestion
一方、独自エンジン系をCSSモジュールのサポートという観点から見ると、厳しい見方をせざるを得ない。EPUBリーダーは、現行規格が求める “CSS Snapshot” はもちろん、前バージョンであるEPUB 3.0.1が求めるCSSモジュールの実装要件 “CSS Profile”（[原文](https://idpf.org/epub/301/spec/epub-contentdocs-20140626.html#sec-css-profile)、[和訳](https://imagedrive.github.io/spec/epub301-contentdocs.xhtml#sec-css)）もサポートしているか怪しい。というのは、<a href="#2-2-3-メジャーなepubリーダーでサポートされるcssモジュール">2-2-3</a>で説明したように、いくつかの独自エンジン系のリーダーでCSS レベル2.1のプロパティである[visibility](https://www.w3.org/TR/2011/REC-CSS2-20110607/visufx.html#visibility)などをサポートしていないことが分かったからだ。
```

ogwata committed:
おわりに-最後まで

ogwata committed:
独自エンジン系の定義

ogwata committed:
モダンではないブラウザーのレイアウトエンジン

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
honto-iOS削除

ogwata committed:
「OK」だった。

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
セクション番号の修正

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
セクション番号の修正

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
日本独自の

MurakamiShinyu commented on chap02.md:
```diff
@@ -752,154 +762,94 @@ lang: 'ja'
 本稿では、現在使われている（つまりモダンな<span class="notetext">「モダン」の定義を補足すると、WebにおけるCSS仕様の使われ方が大きく変わったのは、2014年〜2016年に実装がはじまった[CSS変数（カスタムプロパティ）](https://developer.mozilla.org/ja/docs/Web/CSS/var#%E3%83%96%E3%83%A9%E3%82%A6%E3%82%B6%E3%83%BC%E3%81%AE%E4%BA%92%E6%8F%9B%E6%80%A7)だ。そこでこれが使えるかどうかを重要な指標とした。</span>）ブラウザーのレイアウトエンジンを利用したEPUBリーダーを、一括して**「モダンブラウザー系」**と呼ぶことにする。
 
-他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。これを嫌ってだろうか、コストをかけても独自のレイアウトエンジンを実装するEPUBリーダーも多い。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶ。
+他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。それ以外にも、過去の実装との互換性など、ベンダー固有の事情により独自のレイアウトエンジンを実装するEPUBリーダーもある。またブラウザーのレイアウトエンジンを実装したものの、時間が過ぎて「モダン」ではなくなったものもあるようだ。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶことにする。
```
「ブラウザーのレイアウトエンジンを実装」の「実装」は、別の言葉にしてください。「実装」という語はソフトウェア業界では、実際にプログラムのコードを書いて機能を実現することを言います。既存のパーツを利用するのをパーツを実装するとは言いません（「実装する」ではそのパーツの開発も行うという意味になります）。

```suggestion
他方、レイアウトエンジンは中枢となるモジュールであり、これを外部から調達すれば独自機能の追加がむずかしくなるなど、開発の柔軟性を損なうデメリットが考えられる。それ以外にも、過去の実装との互換性など、ベンダー固有の事情により独自のレイアウトエンジンを実装するEPUBリーダーもある。またブラウザーのレイアウトエンジンを利用して開発したものの、時間が過ぎて「モダン」ではなくなったものもあるようだ。本稿ではこれらを一括して**「独自エンジン系」**と呼ぶことにする。
```

MurakamiShinyu commented:
「日本独自の「EPUBのガラパゴス化」」と書かれていますが、EPUBで利用できるCSS機能が限られるのは日本独自というわけではないようです。「Flexbox in ePub3」という話題があるのを見つけました。

[Flexbox in ePub3 - MobileRead Forums](https://www.mobileread.com/forums/showthread.php?t=338796)

"To what extent is the CSS flexbox layout property supported in ePub 3 e-book reader devices? Is it OK to use it? ePub 3 is supposed to support HTML 5 & CSS 3, right?"
（CSS flexboxレイアウトプロパティは、ePub 3電子書籍リーダーデバイスでどの程度サポートされていますか？　使っても大丈夫ですか？　ePub 3はHTML 5とCSS 3に対応しているはずですよね？）

という質問に対して、

"Epub3 supports subsets of HTML5 and CSS3. As far as I've seen, IRL support for Flexbox is, at best, spotty so be prepared to work around that to produce an epub3 that is compatible with more renderers."
（Epub3はHTML5とCSS3のサブセットをサポートしている。私が見た限りでは、実際のところ、Flexboxのサポートはせいぜいわずかなので、より多くのレンダラーと互換性のあるepub3を作成するために、それを回避する準備をしてください。）

という回答がされています。

ogwata commented:
> 「日本独自の「EPUBのガラパゴス化」」と書かれていますが、EPUBで利用できるCSS機能が限られるのは日本独自というわけではないようです。「Flexbox in ePub3」という話題があるのを見つけました。
> 
了解です。
「日本独自の」をとりましょう。

ogwata committed:
実装→利用、日本独自のをトル

ogwata commented:
ところで、「現行規格が電書協ガイドver.1.1.3を適合とした」という言い方で正しいのでしょうか？
うまい表現がなく苦し紛れで考えたもので、気になっています。

ogwata committed:
推敲

ogwata committed:
推敲

ogwata committed:
推敲

ogwata committed:
過去との互換性も大事

ogwata committed:
ミニマムなCSSのサポートを続ければ

ogwata committed:
テストファイルの開発方針として不適切を削除

ogwata committed:
これをテスト項目から外していたを復活

ogwata committed:
サポートに留まれば

ogwata committed:
ミニマムな

ogwata committed:
熾烈なブラウザーの開発競争

ogwata committed:
夢物語

ogwata committed:
非現実的な話し

ogwata committed:
CSS SnapshotはMUSTと再修正2-1

MurakamiShinyu commented on chap02.md:
```diff
@@ -39,19 +39,46 @@ lang: 'ja'
 3. [大まかな相互運用性のあるCSSモジュール（Modules with Rough Interoperability）](https://www.w3.org/TR/CSS/#rough-interop)
 
-前掲EPUB 3.3の引用にある「CSS Snapshotで定義されているCSS」とは、上記のうち1のことだ。つまりEPUB 3.3では数多くあるCSSモジュールの中でも、上記1でリストアップされているCSSモジュールだけは使えるよう求められている。では、**現在日本国内で利用されているEPUBリーダーは、これらのCSSモジュールをサポートしているのだろうか？**
+前掲EPUB 3.3の引用にある「CSS Snapshotで定義されているCSS」とは上記のうち1を指しており、このサブセットを「サポートする」と宣言している訳だ。さらに同仕様の[6.3 Cascading Style Sheets (CSS)](https://www.w3.org/TR/epub-rs-33/#sec-css)まで読み進むと、“CSS Snapshot” について以下のように規定している。
```
「同仕様の[6.3 Cascading Style Sheets (CSS)]」では、同じEPUB 3.3仕様書のようですが、別の仕様書に分かれているので「EPUB Reading Systems 3.3仕様の」を入れる必要があります。（セクション番号が「6.3」まで共通なのは偶然です）


```suggestion
前掲EPUB 3.3の引用にある「CSS Snapshotで定義されているCSS」とは上記のうち1を指しており、このサブセットを「サポートする」と宣言している訳だ。さらにEPUB Reading Systems 3.3仕様の[6.3 Cascading Style Sheets (CSS)](https://www.w3.org/TR/epub-rs-33/#sec-css)を読むと、“CSS Snapshot” について以下のように規定している。
```

ogwata committed:
EPUB Reading Systems 3.3仕様の

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

MurakamiShinyu commented on chap02.md:
```diff
@@ -420,7 +457,7 @@ lang: 'ja'
       <td bgcolor="#c9daf8" align="center">2024/01/24</td>
     </tr>
     <tr>
```
honto Androidアプリを掲載

```suggestion
    <tr>
      <td bgcolor="#c9daf8" align="center">honto電子書籍リーダー *¹</td>
      <td bgcolor="#c9daf8" align="center">honto</td>
      <td bgcolor="#c9daf8" align="center">Androidアプリ</td>
      <td bgcolor="#c9daf8" align="center">OPPO A55s 5G</td>
      <td bgcolor="#c9daf8" align="center">Android 12</td>
      <td bgcolor="#c9daf8" align="center">6.62.0</td>
      <td bgcolor="#c9daf8" align="center">村上</td>
      <td bgcolor="#c9daf8" align="center">2024/04/01</td>
    </tr>
    <tr>
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -167,11 +194,11 @@ lang: 'ja'
 ### 2-1-3 本調査におけるテスト環境の一覧
 
-今回の調査でテストしたEPUBリーダーの数は、全部で54にのぼる。そのうち、テスト結果を掲載したリーダーは36だ。その明細を<a href="#2-1-3-1-掲載したテスト環境の一覧">2-1-3-1 掲載したテスト環境の一覧</a>に掲げる。
+今回の調査でテストしたEPUBリーダーの数は、テスト環境の違いも入れると、全部で55にのぼる。そのうち、テスト結果を掲載したものは37だ。その明細を<a href="#2-1-3-1-掲載したテスト環境の一覧">2-1-3-1 掲載したテスト環境の一覧</a>に掲げる。
```
honto Android版を追加したので

```suggestion
今回の調査でテストしたEPUBリーダーの数は、テスト環境の違いも入れると、全部で56にのぼる。そのうち、テスト結果を掲載したものは38だ。その明細を<a href="#2-1-3-1-掲載したテスト環境の一覧">2-1-3-1 掲載したテスト環境の一覧</a>に掲げる。
```

ogwata committed:
2-3おわりに-最後まで

ogwata committed:
全部で54にのぼる

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
honto Androidアプリを掲載

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
次いで接頭辞つきを併記

ogwata committed:
日本の電子書籍市場

ogwata commented:
ここまで、必要な修正はしたつもりです。
私の方は、レビュー対応と並行して、これからスライドにとりかかります。

ogwata committed:
仮説は証明された

ogwata committed:
仮説は証明された-削除

MurakamiShinyu commented on chap02.md:
```diff
@@ -29,29 +29,56 @@ lang: 'ja'
+> 和訳：**§1.3.3 CSSとの関係**　EPUB3は、CSS Working Group Snapshot [[csssnapshot]](https://www.w3.org/TR/CSS/)で定義されているCSSをサポートする。EPUB3はグローバル言語のための一貫したサポートを保証するために、いくつかの接頭辞付きのCSSプロパティも維持する。
 
-上記にある “CSS Working Group Snapshot”（以下、“CSS Snapshot”）とはなんだろう。CSSレベル2まで、CSS仕様はすべてのプロパティを単一の仕様書で規定していた。しかしレベル2を超えるCSSでは飛躍的に規模が大きくなることが確実になった。そこで審議するCSS Working Groupは、機能や目的ごとの「モジュール」に仕様を分割することにし、更新はモジュールごとにすすめることにした。ところが更新のペースがバラバラになると、今度はCSS仕様全体の状況が把握しづらくなってしまう。そこで考え出されたのが、Webブラウザー（以下、ブラウザー）に実装されたCSSモジュールを、安定性の程度を評価・分類したリストを作成し、これを1年ごとに更新していく “CSS Snapshot” だ。<span class="notetext"> “CSS Snapshot” ...
+上記にある “CSS Working Group Snapshot”（以下、“CSS Snapshot”）とはなんだろう。CSSレベル2まで、CSS仕様はすべてのプロパティを単一の仕様書で規定していた。しかしレベル2を超えるCSSでは飛躍的に規模が大きくなることが確実になった。そこで審議するCSS Working Groupは、機能や目的ごとの「モジュール」に仕様を分割することにし、更新はモジュールごとにすすめることにした。ところが更新のペースがバラバラになると、今度はCSS仕様全体の状況が把握しづらくなってしまう。そこで考え出されたのが、Webブラウザー（以下、ブラウザー）に実装されたCSSモジュールを、安定性（普及の進み具合）の程度を評価・分類したリストを作成し、これを1年ごとに更新していく “CSS Snapshot” だ。<span class="notetext"> “CSS  ...
```
「安定性（普及の進み具合）」は間違いです。そのことは https://www.w3.org/TR/css/#css-official のNOTEに次のように書かれています。

> NOTE: …: inclusion in this profile is based on feature stability only and not on expected use or Web browser adoption.
（このプロファイルに含まれるのは、機能の安定性だけであり、予想される使用状況やウェブブラウザでの採用状況には基づいていない。）

```suggestion
上記にある “CSS Working Group Snapshot”（以下、“CSS Snapshot”）とはなんだろう。CSSレベル2まで、CSS仕様はすべてのプロパティを単一の仕様書で規定していた。しかしレベル2を超えるCSSでは飛躍的に規模が大きくなることが確実になった。そこで審議するCSS Working Groupは、機能や目的ごとの「モジュール」に仕様を分割することにし、更新はモジュールごとにすすめることにした。ところが更新のペースがバラバラになると、今度はCSS仕様全体の状況が把握しづらくなってしまう。そこで考え出されたのが、Webブラウザー（以下、ブラウザー）に実装されたCSSモジュールを、安定性の程度を評価・分類したリストを作成し、これを1年ごとに更新していく “CSS Snapshot” だ。<span class="notetext"> “CSS Snapshot” のURLは<https://www.w3.org/TR/CSS/>に固定されている。本稿を執筆している2024年3月では “CSS Snapshot 2023” というタイトルだが、1年後の2025年3月に同じURLをクリックすれば “CSS Snapshot 2024” になっているはずだ。なお、2023年版のパーマネント・リンクは<https://www.w3.org/TR/css-2023/>である。</span>
```

ogwata committed:
安定性の程度を評価

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2620,113 @@ lang: 'ja'
+
+その上での仮説だが、独自エンジン系のリーダーがミニマムな「20150101-CSS機能一覧.pdf」のサポートだけでよいと割り切った実装をしたとすれば、最新のCSSモジュールのサポートに消極的だった理由として説明がつくのではないか。
+
+たとえば、<a href="#2-2-4-2writing-mode-プロパティ-vertical-rlに絞ったテスト結果">2-1-2</a>、及び<a href="#2-2-1-レイアウトエンジンと本テストとの関係">2-2-4-2</a>で述べたように、Kindle（Windows版）、kobo-other (Mac版／Windows版)、超縦書の4つはwriting-mode プロパティは接頭辞をつける必要がある。これはまさに電書協ガイドver.1.1.3の通りなのだ。では、これら以外の独自エンジン系リーダーは、なぜ接頭辞をつけないよう修正したのかといえば、仕様が勧告になったのでその必要がなくなったと考えるのが自然だろう。<span class="notetext">むしろ勧告後も接頭辞を付けるルールを変えない電書協ガイドの方が、W3C仕様から見て不自然であることは指摘したい。互 ...
+
```
writing-modeのテスト結果が、「独自エンジン系のリーダーがミニマムな「20150101-CSS機能一覧.pdf」のサポートだけでよいと割り切った実装をした」ことを示しているかのように書かれてますが、間違っていると思います。

kobo-otherと超縦書が接頭辞を必要としている理由は、電書協ガイドとは関係なく、これらは実際は独自エンジンではなく、それぞれ2013年ごろのWebKit、2015年ごろのChromiumを使っており、その当時のそれらのブラウザーでは接頭辞を必要としていたためです。

残るKindle (Windows版)ですが、Windows以外ではアップデートで接頭辞を必要としなくなっていることっから、Windows版は開発された時期が古くてそれからCSSサポートについてアップデートされないままになっていると考えられます。

Kindle (Windows版)を除くとすべての本当に（古いブラウザではなく）独自エンジンを使うリーダーで、標準仕様の（接頭辞なしの）writing-modeがサポートされています。

この結果から、「独自エンジン系のリーダーがミニマムな「20150101-CSS機能一覧.pdf」のサポートだけでよいと割り切った実装をした」こととはまったく結びつきません。

この段落は丸ごと削除したほうがよいでしょう。

```suggestion
```

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2620,113 @@ lang: 'ja'
+
+では独自エンジン系のリーダーが、最新のCSSモジュールではなくミニマムな電書協ガイドのサポートに留まっている可能性が高いことを、私達はどのように考えればよいだろう。
+
+たとえば、Webページで書きためた自分の原稿を、電子書籍として売ろうと考えた人がいたとする。なぜならその人は、有り物のHTMLとCSSを最少限の手間で電子書籍にすることを、EPUB 3.3が可能にしたことを知っていたからだ。ところが、そうしてできたEPUBを[Kindleダイレクト・パブリッシング](https://kdp.amazon.co.jp/ja_JP/)で自主出版しようとしても、EPUB 3.3には適合するはずなのに、肝腎のKindleのリーダーでは意図通りに表示されない可能性が高い。なぜなら、現在のWebページでは普通に使えるCSS変数やCSS Flexible Box Module Level 1、あるいはCSS Counter Styles Level 3に、Kindleは対応していないからだ。
```
「CSS変数やCSS Flexible Box Module Level 1、あるいはCSS Counter Styles Level 3」は「CSS変数やCSS Flexible Box」だけのほうがよいと思います。
まず、CSS変数のほうは仕様の通称であるのに、「CSS Flexible Box Module Level 1」は長い名前なのはバランスがとれません。「CSS Counter Styles Level 3」はあまり活用されているのを見たことがありません。「CSS変数やCSS Flexible Box」だけのほうが「現在のWebページでは普通に使える」ものとしてわかりやすいと思います。


```suggestion
たとえば、Webページで書きためた自分の原稿を、電子書籍として売ろうと考えた人がいたとする。なぜならその人は、有り物のHTMLとCSSを最少限の手間で電子書籍にすることを、EPUB 3.3が可能にしたことを知っていたからだ。ところが、そうしてできたEPUBを[Kindleダイレクト・パブリッシング](https://kdp.amazon.co.jp/ja_JP/)で自主出版しようとしても、EPUB 3.3には適合するはずなのに、肝腎のKindleのリーダーでは意図通りに表示されない可能性が高い。なぜなら、現在のWebページでは普通に使えるCSS変数やCSS Flexible Boxに、Kindleは対応していないからだ。
```

ogwata committed:
電書協ガイドver.1.1.3の通り-削除

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

ogwata committed:
CSS変数やCSS Flexible Box

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2620,113 @@ lang: 'ja'
+
+たとえば、Webページで書きためた自分の原稿を、電子書籍として売ろうと考えた人がいたとする。なぜならその人は、有り物のHTMLとCSSを最少限の手間で電子書籍にすることを、EPUB 3.3が可能にしたことを知っていたからだ。ところが、そうしてできたEPUBを[Kindleダイレクト・パブリッシング](https://kdp.amazon.co.jp/ja_JP/)で自主出版しようとしても、EPUB 3.3には適合するはずなのに、肝腎のKindleのリーダーでは意図通りに表示されない可能性が高い。なぜなら、現在のWebページでは普通に使えるCSS変数やCSS Flexible Box Module Level 1、あるいはCSS Counter Styles Level 3に、Kindleは対応していないからだ。
+
+EPUB 3.3で “CSS Snapshot” のサポートが必須になった理由は、なにもベンダーに嫌がらせをするためではない。現在のWebページをそのまま電子書籍として流通できるようにすることで、Webによるコミュニケーションをステップアップさせる狙いがあったからではないか。つまり、Webページと電子書籍の統合である。
```
「EPUB 3.3で “CSS Snapshot” のサポートが必須になった」では、それまでのEPUBバージョンでは必須でなかったものが急に必須になったかのようです。2017年1月のEPUB3.1から必須だし、その前の2014年のEPUB 3.0.1のころは[CSS Snapshot 2015](https://www.w3.org/TR/css-2015/)ができる前だったので、それの代わりに当時安定していたCSS仕様をEPUB3 CSS Profileとして取り込んでいたと考えると、EPUB仕様でCSSサポートの方針がとくに変わったわけではなくて、CSS Snapshotができたことで、EPUB3 CSS Profileが不要となっただけといえます。

「必須になった理由」を「必須とされている理由」に直すとよいと思います。

EPUB仕様がCSSサポートの方針を変えたわけではないのだから、「なにもベンダーに嫌がらせをするためではない」も余計でしょう。

「現在のWebページをそのまま電子書籍として流通できるようにすること」
はちょっと無理があります。EPUBのコンテンツはまだWebでは推奨されなくなったXHTMLです。
「Webページをそのまま」を「Webページをもとにして」などと直すとよいかもしれません。

ogwata replied:
以下のように直しました。

> EPUB 3.3で “CSS Snapshot” のサポートが必須とされた理由は、現在のWebページを元に少ない手間で電子書籍として流通できるようにすることで、Webによるコミュニケーションをステップアップさせる狙いがあったからではないか。つまり、Webページと電子書籍の統合である。

MurakamiShinyu commented on chap02.md:
```diff
@@ -2736,16 +2620,111 @@ lang: 'ja'
+
+しかしEPUB 3.3の狙いと独自エンジン系リーダーとの間には、きわめて大きな断絶があると言わざるを得ない。このまま日本のEPUBリーダーがサポートするCSSが、ずっと電書協ガイドの範囲に限定され続ければ今後どうなるだろう？　本テストの結果を見ながら、そのような疑問にとらわれることも確かだ。
+
+もちろん過去との互換性も大事かもしれない。とはいえ、ミニマムなCSSのサポートに留まることで、日本の電子書籍市場にバラ色の未来がやってくるのかと聞かれれば、誰もが首を振るしかないのも確かではないだろうか。
+
+
```
この最後の段落は削除して、前の段落で終わりにするのでよいと思います。削除したい理由は：

- 「誰もが首を振るしかない」という慣用表現ですが、横に振るのか縦に振るのかどっちなのだろうと混乱する人もいます。
- 電子書籍でWebと同じCSSが使えることに期待しているような人は限られているでしょう。「誰もが……確か」ではないでしょう。
- あまり最後まで書かないで、読み手に考えさせるような終わり方でよいと思います。


```suggestion
```

ogwata committed:
サポートが必須とされた

ogwata committed:
最終段落削除

Co-authored-by: Shinyu Murakami <murakami@vivliostyle.org>
